### PR TITLE
dev-qt/qtwebengine: add check-reqs for disk space and memory usage

### PR DESCRIPTION
**Pull request by thesamesam**
Not really married to any of the implementation details here. Just wanted to get a draft up so we can discuss it. Comment away!

---

For now, assumes:
- RAM/memory: 2GB per job for GCC, 1.5GB per job for Clang
- Disk (build): 7GB
- Disk (installed): none yet

(Also moves pkg_preinst to the end of the ebuild to reflect
phase execution order, as per devmanual.)

Closes: https://bugs.gentoo.org/570534
Signed-off-by: Sam James <sam@gentoo.org>

### CODE REVIEW

**Comment by thesamesam** [comment]
[02:43:51]  <@sam_> have wondered about just adding something to check-reqs or multiprocessing.eclass to just clamp MAKEOPTS jobs for big packages

**Review by ionenwks — COMMENTED** [review]
fwiw I build this with clang+thinlto+jumbo, my build for this snapshot used 5GB disk so 7GB sound reasonable for that too (no point in making these checks too tight imo).

Also used -j11 but it never came close to using 16GB (not that I paid attention to clang processes, but it's better to be on the safe side here too so 1.5 is probably fair)

File: dev-qt/qtwebengine/qtwebengine-5.15.2_p20211210.ebuild [hunk 1]
```diff
@@ -110,15 +110,39 @@ PATCHES=(
...
-pkg_preinst() {
-	elog "This version of Qt WebEngine is based on Chromium version 87.0.4280, with"
-	elog "additional security fixes from newer versions. Extensive as it is, the"
-	elog "list of backports is impossible to evaluate, but always bound to be behind"
-	elog "Chromium's release schedule."
-	elog "In addition, various online services may deny service based on an outdated"
-	elog "user agent version (and/or other checks). Google is already known to do so."
-	elog
-	elog "tldr: Your web browsing experience will be compromised."
+CHECKREQS_DISK_BUILD="7G"
```
[the remark this answers is not in the document]
Comment (thesamesam) [reply dev-qt/qtwebengine/qtwebengine-5.15.2_p20211210.ebuild:113]: I'd seen that before (a while ago) and thought it was odd. Apparently totally forgot this time.

File: dev-qt/qtwebengine/qtwebengine-5.15.2_p20211210.ebuild [hunk 2]
```diff
@@ -110,15 +110,39 @@ PATCHES=(
...
+
+	# Let's crudely assume ~2GB per compiler job for GCC.
+	local multiplier=2
+
+	# And call it ~1.5GB for Clang.
+	if tc-is-clang ; then
+		multiplier=1.5
+	fi
+
+	echo "$(($(makeopts_jobs)*${multiplier}))M"
```
[the remark this answers is not in the document]
Comment (thesamesam) [reply dev-qt/qtwebengine/qtwebengine-5.15.2_p20211210.ebuild]: Thanks, I should've realised the FP thing (you've probably said it before, thinking about it).

File: dev-qt/qtwebengine/qtwebengine-5.15.2_p20211210.ebuild [hunk 3]
```diff
@@ -110,15 +110,39 @@ PATCHES=(
...
+}
+
+pkg_pretend() {
+	if [[ ${MERGE_TYPE} != binary ]] ; then
+		# check-reqs has a binary guard but let's wrap the whole thing
+		# for the memory-estimator so we don't try find e.g. CC on binpkg consumers.
+		export CHECKREQS_MEMORY=$(_required_ram_available)
```
[the remark this answers is not in the document]
Comment (thesamesam) [reply dev-qt/qtwebengine/qtwebengine-5.15.2_p20211210.ebuild]: Very snazzy!

File: dev-qt/qtwebengine/qtwebengine-5.15.2_p20211210.ebuild [hunk 4]
```diff
@@ -110,29 +110,47 @@ PATCHES=(
...
 	eshopts_push -s extglob
 	if is-flagq '-g?(gdb)?([1-9])'; then
-		ewarn
-		ewarn "You have enabled debug info (probably have -g or -ggdb in your CFLAGS/CXXFLAGS)."
-		ewarn "You may experience really long compilation times and/or increased memory usage."
-		ewarn "If compilation fails, please try removing -g/-ggdb before reporting a bug."
-		ewarn
+		# Debug symbols require significantly more memory for the compiler
+		# to juggle.
+		multiplier=$((multiplier*2))
```
Comment (Pesa) [inline dev-qt/qtwebengine/qtwebengine-5.15.2_p20211210.ebuild:133]: doesn't it require more disk space as well?

File: dev-qt/qtwebengine/qtwebengine-5.15.2_p20211210.ebuild [hunk 5]
```diff
@@ -110,29 +110,47 @@ PATCHES=(
...
 	fi
 	eshopts_pop
 
+	local CHECKREQS_DISK_BUILD="7G"
+	local CHECKREQS_DISK_USR="7G"
+	local CHECKREQS_MEMORY=$(($(makeopts_jobs)*${multiplier}/10))G
```
Comment (Pesa) [inline dev-qt/qtwebengine/qtwebengine-5.15.2_p20211210.ebuild:139]: I believe the `${}` is redundant inside `$((...))`

File: dev-qt/qtwebengine/qtwebengine-5.15.2_p20211210.ebuild [hunk 6]
```diff
@@ -110,29 +110,47 @@ PATCHES=(
...
+		multiplier=15
+	fi
+
+	# bug #307861
 	eshopts_push -s extglob
 	if is-flagq '-g?(gdb)?([1-9])'; then
-		ewarn
-		ewarn "You have enabled debug info (probably have -g or -ggdb in your CFLAGS/CXXFLAGS)."
-		ewarn "You may experience really long compilation times and/or increased memory usage."
-		ewarn "If compilation fails, please try removing -g/-ggdb before reporting a bug."
```
Comment (Pesa) [inline dev-qt/qtwebengine/qtwebengine-5.15.2_p20211210.ebuild:131]: hmm I would keep the warning... maybe? Not a strong reason, but it has happened before that users "forgot" that they have debugging flags in their CXXFLAGS... On the other hand, who reads the warnings anyway? 😛

File: dev-qt/qtwebengine/qtwebengine-5.15.2_p20211210.ebuild (hunk 5 above)
Comment (thesamesam) [reply dev-qt/qtwebengine/qtwebengine-5.15.2_p20211210.ebuild:139]: I think you're right. Ionen mentioned it [here](https://github.com/gentoo/gentoo/pull/23310#discussion_r769216237). It's not my ebuild so I'm happy to use no ${} if you prefer that.

File: dev-qt/qtwebengine/qtwebengine-5.15.2_p20211210.ebuild (hunk 4 above)
Comment (thesamesam) [reply dev-qt/qtwebengine/qtwebengine-5.15.2_p20211210.ebuild:133]: Ah, yes!

File: dev-qt/qtwebengine/qtwebengine-5.15.2_p20211210.ebuild (hunk 6 above)
Comment (thesamesam) [reply dev-qt/qtwebengine/qtwebengine-5.15.2_p20211210.ebuild:131]: Given the aim here is to hopefully reduce the # of people reporting OOM bugs, I think you're right here, I'll keep the warning?

File: dev-qt/qtwebengine/qtwebengine-5.15.2_p20211210.ebuild (hunk 4 above)
Comment (ionenwks) [reply dev-qt/qtwebengine/qtwebengine-5.15.2_p20211210.ebuild:133]: No sure if debugging really need that much more memory (would need to check), when I mentioned x2 it was mostly for disk space.

Comment (ionenwks) [reply dev-qt/qtwebengine/qtwebengine-5.15.2_p20211210.ebuild:133]: It may be more than x2 too, I think chromium can get absurd with this.

Comment (thesamesam) [reply dev-qt/qtwebengine/qtwebengine-5.15.2_p20211210.ebuild:133]: Sorry, yes, that's obvious now you say it. I don't have a good estimate for how much more _memory_ usage it'll cause.

Might be worth not addressing it for memory and just handling it for disk where we can at least say >= 2x safely, I think.

File: dev-qt/qtwebengine/qtwebengine-5.15.2_p20211210.ebuild (hunk 6 above)
Comment (thesamesam) [reply dev-qt/qtwebengine/qtwebengine-5.15.2_p20211210.ebuild]: OK, kept it. And dropped the logic for messing with check-reqs sizes based on debugging symbols for now for simplicity.

File: dev-qt/qtwebengine/qtwebengine-5.15.2_p20211210.ebuild (hunk 4 above)
Comment (thesamesam) [reply dev-qt/qtwebengine/qtwebengine-5.15.2_p20211210.ebuild]: https://github.com/gentoo/gentoo/pull/23310#discussion_r771086219.

**Comment by a17r** [comment]
sure why not.

File: dev-qt/qtwebengine/qtwebengine-5.15.2_p20211210.ebuild [hunk 7]
```diff
@@ -110,29 +110,47 @@ PATCHES=(
...
+	# Let's crudely assume ~2GB per compiler job for GCC.
+	local multiplier=20
+
+	# And call it ~1.5GB for Clang.
+	if tc-is-clang ; then
+		multiplier=15
+	fi
+
+	local CHECKREQS_DISK_BUILD="7G"
+	local CHECKREQS_DISK_USR="7G"
```
Comment (Pesa) [inline dev-qt/qtwebengine/qtwebengine-5.15.2_p20211210.ebuild:138]: uhm really 7 GB of installed size? I don't currently have qtwebengine installed so cannot quickly check... can anyone confirm?

**Review by Pesa — COMMENTED** [review]
Looks good to me, though you may want to double check that value for `CHECKREQS_DISK_USR`, seems too high to me.

File: dev-qt/qtwebengine/qtwebengine-5.15.2_p20211210.ebuild (hunk 7 above)
Comment (a17r) [reply dev-qt/qtwebengine/qtwebengine-5.15.2_p20211210.ebuild:138]:
```
# equery s qtwebengine
 * dev-qt/qtwebengine-5.15.2_p20211216
         Total files : 263
         Total size  : 149.60 MiB
```

Comment (thesamesam) [reply dev-qt/qtwebengine/qtwebengine-5.15.2_p20211210.ebuild:138]: yeah, seems obviously wrong, will fix

**Comment by thesamesam** [comment]
Done in 2b134e09a3ab94f3c33419cf0b8114d004cec477 thanks all!

**Pull request closed without merging**